File: backend/src/services/ai/modelPricing.js
const DEFAULT_CURRENCY = "USD";

// Default rates are USD per 1M tokens.
// Verified on April 22, 2026 from official provider pricing pages.
const DEFAULT_MODEL_PRICING = {
  openai: {
    "gpt-5.4": { input: 2.5, output: 15.0 },
    "gpt-5.4-mini": { input: 0.75, output: 4.5 },
    "gpt-5.4-nano": { input: 0.2, output: 1.25 },
    "gpt-5": { input: 1.25, output: 10.0 },
    "gpt-5-mini": { input: 0.25, output: 2.0 },
    "gpt-5-nano": { input: 0.05, output: 0.4 },
  },
  gemini: {
    "gemini-2.5-flash": { input: 0.3, output: 2.5 },
    "gemini-2.5-flash-lite": { input: 0.1, output: 0.4 },
  },
};

function normalizeProvider(value) {
  return String(value || "").trim().toLowerCase();
}

function normalizeModel(value) {
  return String(value || "").trim().toLowerCase();
}

function cloneCatalog(catalog) {
  const cloned = {};
  for (const [provider, models] of Object.entries(catalog || {})) {
    cloned[provider] = {};
    for (const [model, spec] of Object.entries(models || {})) {
      cloned[provider][model] = { ...spec };
    }
  }
  return cloned;
}

function coerceSpec(raw) {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;

  const inputRate =
    raw.input ?? raw.input_per_million ?? raw.input_price_per_million;
  const outputRate =
    raw.output ?? raw.output_per_million ?? raw.output_price_per_million;

  const input = Number(inputRate);
  const output = Number(outputRate);
  if (!Number.isFinite(input) || !Number.isFinite(output)) return null;

  return {
    input: Math.max(0, input),
    output: Math.max(0, output),
    currency: String(raw.currency || DEFAULT_CURRENCY).toUpperCase(),
  };
}

let cachedCatalog = null;

function getPricingCatalog() {
  if (cachedCatalog) return cachedCatalog;

  const catalog = cloneCatalog(DEFAULT_MODEL_PRICING);
  const raw = String(process.env.AI_MODEL_PRICING_JSON || "").trim();
  if (!raw) {
    cachedCatalog = catalog;
    return catalog;
  }

  try {
    const parsed = JSON.parse(raw);
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      for (const [provider, models] of Object.entries(parsed)) {
        const providerKey = normalizeProvider(provider);
        if (!providerKey || !models || typeof models !== "object" || Array.isArray(models)) {
          continue;
        }

        catalog[providerKey] = catalog[providerKey] || {};
        for (const [model, rawSpec] of Object.entries(models)) {
          const spec = coerceSpec(rawSpec);
          if (!spec) continue;
          catalog[providerKey][normalizeModel(model)] = {
            ...spec,
            _source: "env:AI_MODEL_PRICING_JSON",
          };
        }
      }
    }
  } catch {
    // Ignore malformed override and fall back to built-in defaults.
  }

  cachedCatalog = catalog;
  return catalog;
}

export function resolvePricing({ provider, model }) {
  const modelKey = normalizeModel(model);
  if (!modelKey) return null;

  const providerKey = normalizeProvider(provider);
  const catalog = getPricingCatalog();
  const candidateProviders = providerKey ? [providerKey] : Object.keys(catalog);

  for (const currentProvider of candidateProviders) {
    const providerCatalog = catalog[currentProvider] || {};
    if (providerCatalog[modelKey]) {
      const spec = providerCatalog[modelKey];
      return {
        ...spec,
        provider: currentProvider,
        canonicalModel: modelKey,
        source: spec._source || "default_catalog",
      };
    }

    const candidateModels = Object.keys(providerCatalog).sort(
      (a, b) => b.length - a.length,
    );
    for (const candidateModel of candidateModels) {
      if (modelKey.startsWith(candidateModel)) {
        const spec = providerCatalog[candidateModel];
        return {
          ...spec,
          provider: currentProvider,
          canonicalModel: candidateModel,
          source: spec._source || "default_catalog",
        };
      }
    }
  }

  return null;
}

export function estimateUsageCost({
  provider,
  model,
  inputTokens = 0,
  outputTokens = 0,
}) {
  const pricing = resolvePricing({ provider, model });
  if (!pricing) return null;

  const inTok = Math.max(0, Number(inputTokens || 0));
  const outTok = Math.max(0, Number(outputTokens || 0));

  const inputCost = (inTok * pricing.input) / 1_000_000;
  const outputCost = (outTok * pricing.output) / 1_000_000;

  return {
    provider: pricing.provider,
    canonicalModel: pricing.canonicalModel,
    currency: pricing.currency || DEFAULT_CURRENCY,
    inputCost,
    outputCost,
    totalCost: inputCost + outputCost,
    source: pricing.source || "default_catalog",
  };
}

File: services/analytics/model_pricing.py
from __future__ import annotations

import json
import os
from functools import lru_cache
from typing import Any, Dict, Optional

DEFAULT_CURRENCY = "USD"

# Default rates are in USD per 1M tokens.
# Verified on April 22, 2026 from official pricing pages:
# - OpenAI API pricing: gpt-5.4-nano input $0.20 / output $1.25 per 1M tokens
# - Gemini Developer API pricing: gemini-2.5-flash standard input $0.30 / output $2.50 per 1M tokens
DEFAULT_MODEL_PRICING: Dict[str, Dict[str, Dict[str, Any]]] = {
    "openai": {
        "gpt-5.4": {"input": 2.50, "output": 15.00},
        "gpt-5.4-mini": {"input": 0.75, "output": 4.50},
        "gpt-5.4-nano": {"input": 0.20, "output": 1.25},
        "gpt-5": {"input": 1.25, "output": 10.00},
        "gpt-5-mini": {"input": 0.25, "output": 2.00},
        "gpt-5-nano": {"input": 0.05, "output": 0.40},
    },
    "gemini": {
        "gemini-2.5-flash": {"input": 0.30, "output": 2.50},
        "gemini-2.5-flash-lite": {"input": 0.10, "output": 0.40},
    },
}


def _normalize_provider(value: Optional[str]) -> str:
    return str(value or "").strip().lower()


def _normalize_model(value: Optional[str]) -> str:
    return str(value or "").strip().lower()


def _clone_catalog(catalog: Dict[str, Dict[str, Dict[str, Any]]]) -> Dict[str, Dict[str, Dict[str, Any]]]:
    return {
        provider: {
            model: dict(spec)
            for model, spec in models.items()
        }
        for provider, models in catalog.items()
    }


def _coerce_spec(raw: Any) -> Optional[Dict[str, Any]]:
    if not isinstance(raw, dict):
        return None

    input_rate = raw.get("input")
    if input_rate is None:
        input_rate = raw.get("input_per_million")
    if input_rate is None:
        input_rate = raw.get("input_price_per_million")

    output_rate = raw.get("output")
    if output_rate is None:
        output_rate = raw.get("output_per_million")
    if output_rate is None:
        output_rate = raw.get("output_price_per_million")

    try:
        input_rate = float(input_rate)
        output_rate = float(output_rate)
    except (TypeError, ValueError):
        return None

    return {
        "input": max(0.0, input_rate),
        "output": max(0.0, output_rate),
        "currency": str(raw.get("currency") or DEFAULT_CURRENCY).upper(),
    }


def _apply_override(
    catalog: Dict[str, Dict[str, Dict[str, Any]]],
    provider: str,
    model: str,
    spec: Dict[str, Any],
    source: str,
) -> None:
    provider_key = _normalize_provider(provider)
    model_key = _normalize_model(model)
    if not provider_key or not model_key:
        return

    catalog.setdefault(provider_key, {})
    merged = dict(spec)
    merged["_source"] = source
    catalog[provider_key][model_key] = merged


@lru_cache(maxsize=1)
def get_pricing_catalog() -> Dict[str, Dict[str, Dict[str, Any]]]:
    catalog = _clone_catalog(DEFAULT_MODEL_PRICING)

    raw = (os.getenv("AI_MODEL_PRICING_JSON") or "").strip()
    if not raw:
        return catalog

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return catalog

    if not isinstance(parsed, dict):
        return catalog

    for provider, models in parsed.items():
        provider_key = _normalize_provider(provider)
        if not provider_key or not isinstance(models, dict):
            continue

        for model, raw_spec in models.items():
            spec = _coerce_spec(raw_spec)
            if spec:
                _apply_override(catalog, provider_key, str(model), spec, "env:AI_MODEL_PRICING_JSON")

    return catalog


def resolve_pricing(provider: Optional[str], model: Optional[str]) -> Optional[Dict[str, Any]]:
    provider_key = _normalize_provider(provider)
    model_key = _normalize_model(model)
    if not model_key:
        return None

    catalog = get_pricing_catalog()
    candidate_providers = [provider_key] if provider_key else list(catalog.keys())

    for provider_name in candidate_providers:
        provider_catalog = catalog.get(provider_name) or {}
        if not provider_catalog:
            continue

        if model_key in provider_catalog:
            spec = dict(provider_catalog[model_key])
            spec["provider"] = provider_name
            spec["canonical_model"] = model_key
            spec["source"] = spec.pop("_source", "default_catalog")
            return spec

        for candidate_model in sorted(provider_catalog.keys(), key=len, reverse=True):
            if model_key.startswith(candidate_model):
                spec = dict(provider_catalog[candidate_model])
                spec["provider"] = provider_name
                spec["canonical_model"] = candidate_model
                spec["source"] = spec.pop("_source", "default_catalog")
                return spec

    return None


def estimate_usage_cost(
    *,
    provider: Optional[str],
    model: Optional[str],
    input_tokens: Optional[int],
    output_tokens: Optional[int],
) -> Optional[Dict[str, Any]]:
    pricing = resolve_pricing(provider, model)
    if pricing is None:
        return None

    try:
        in_tok = max(0, int(input_tokens or 0))
        out_tok = max(0, int(output_tokens or 0))
    except (TypeError, ValueError):
        return None

    input_cost = (in_tok * float(pricing["input"])) / 1_000_000
    output_cost = (out_tok * float(pricing["output"])) / 1_000_000

    return {
        "provider": pricing["provider"],
        "canonical_model": pricing["canonical_model"],
        "currency": pricing.get("currency", DEFAULT_CURRENCY),
        "input_cost": input_cost,
        "output_cost": output_cost,
        "total_cost": input_cost + output_cost,
        "source": pricing.get("source", "default_catalog"),
    }
